sgkit/distance/core.py

```diff
@@ -0,0 +1,163 @@
+    [v2.v0, v2.v1, v2.v2, v2.v3]
+    [v3.v0, v3.v1, v3.v2, v3.v3]
+
+    The (i, j) position in the resulting array (matrix) denotes the distance
+    between vi and vj vectors.
+
+    The given array can be either of type numpy or dask array. For better
+    performance, it is recommended to use dask array with suitable
+    chunking, which makes it highly parallelizable.
+
+    Algorithm: This function works on the philosophy of map/reduce style
+    computation.The map step, which is implemented in the given map_fn
+    is applied on each pair of chunk of a vector, with every chunk of
+    every other vector. Consider the example of the above mentioned array
+
+    The matrix is divided in chunks, lets assume the chunk size is (2, 2).
+    The first chunk would be:
+
+    [e.00, e.01]
+    [e.10, e.11]
+
+    and the next chunk along the y-axis would be:
+
+    [e.20, e.21]
+    [e.30, e.31]
+
+    Now instead of passing the whole vectors: v0, v1, v2, v3 for the
+    calculation of the distance, only chunks of them are passed and the map
+    function calculates partial, between the chunks of two or more vectors and
+    this calculation can happen in parallel, hence speeding up the whole process.
+
+    After the partial calculations of the chunks are done in the map step, they
+    are accumulated in the reduce step, via the given reduce function.
+
+    The importance of ``chunks`` can be seen from the mentioned algorith above,
+    better the chunking better the parallelisation.
+
+    Parameters
+    ----------
+    x
+        [array-like, shape: (M, N)]
+        An array like two dimensional matrix.
+    map_fn
+        Map function for the distance metric.
+    reduce_fn
+        Reduce function for the distance metric
+    n_map_param
+        An integer which represents the dimension of the output.
+    chunks
+        The chunksize for the given array, to be used to convert
+        the given array to dask array if x is of type ndarray.
+
+    Returns
+    -------
+    A two dimensional distance matrix.
+
+    Examples
+    --------
+
+    >>> from sgkit.distance.core import pairwise
+    >>> from sgkit.distance.metrics import correlation_map, correlation_reduce, N_MAP_PARAM
+    >>> import numpy as np
+    >>> import dask.array as da
+    >>> x = da.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]]).rechunk(2, 2)
+    >>> pairwise(x, correlation_map, correlation_reduce, N_MAP_PARAM['correlation']).compute()
+    array([[1.        , 0.73704347, 0.99717646],
+           [0.73704347, 1.        , 0.78571429],
+           [0.99717646, 0.78571429, 1.        ]])
+
+    >>> x = np.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]])
+    >>> pairwise(x, correlation_map, correlation_reduce, N_MAP_PARAM['correlation'], chunks=(2, 2)).compute()
+    array([[1.        , 0.73704347, 0.99717646],
+           [0.73704347, 1.        , 0.78571429],
+           [0.99717646, 0.78571429, 1.        ]])
+
+    """
+    if isinstance(x, np.ndarray) and not chunks:
```

Comment:
FWIW I think it would be fine to just raise an error if the input is not a dask array, rather than trying to be helpful.

Reply:
The approach we've taken in other functions is to say `x = da.asarray(x)` instead.  I think it's safe to assume that if a numpy array was provided, there is no reason to worry about it becoming a single chunk dask array.